src/main/java/ru/vk/itmo/viktorkorotkikh/SSTable.java

```diff
@@ -127,40 +175,121 @@ public static void save(Collection<Entry<MemorySegment>> entries, int fileIndex,
+            // by default file contains JAVA_BYTE == 0 with offset 0
+            // so if we have possibly compacted file and it has JAVA_BYTE == 0 with offset 0
+            // then it is corrupted
+            // otherwise we have unbroken file without tombstones (compacted)
+            // owing to this we use hasNoTombstones condition
+            boolean hasNoTombstones = true;
+            while (entryIterator.hasNext()) {
+                Entry<MemorySegment> entry = entryIterator.next();
+                if (entry.value() == null) {
+                    hasNoTombstones = false;
+                }
                 offset += writeMemorySegment(mappedSSTableFile, entry.key(), offset);
                 offset += writeMemorySegment(mappedSSTableFile, entry.value(), offset);
             }
 
             mappedSSTableFile.force();
-            Files.move(
-                    tmpSSTable,
-                    basePath.resolve(FILE_NAME + fileIndex + FILE_EXTENSION),
-                    StandardCopyOption.ATOMIC_MOVE
+            mappedSSTableFile.set(ValueLayout.JAVA_BOOLEAN, 0, hasNoTombstones);
+            mappedSSTableFile.force();
+            return tmpSSTable;
+        }
+    }
+
+    private static void writeIndex(
+            Iterator<Entry<MemorySegment>> iterator,
+            MemorySegment mappedSSTableFile,
+            long offset
+    ) {
+        long index = 0;
+        while (iterator.hasNext()) {
+            mappedSSTableFile.set(
+                    ValueLayout.JAVA_LONG_UNALIGNED,
+                    METADATA_SIZE + index * ENTRY_METADATA_SIZE,
+                    offset
             );
+            if (iterator instanceof MergeIterator.MergeIteratorWithTombstoneFilter mergeIterator) {
+                offset += mergeIterator.getPointerSizeAndShift();
+            } else {
+                Entry<MemorySegment> entry = iterator.next();
+                offset += Utils.getEntrySize(entry);
+            }
+            index++;
         }
     }
 
-    private static long getEntrySize(Entry<MemorySegment> entry) {
-        if (entry.value() == null) {
-            return Long.BYTES + entry.key().byteSize() + Long.BYTES;
+    public static Path compact(
+            Supplier<MergeIterator.MergeIteratorWithTombstoneFilter> data,
+            Path basePath
+    ) throws IOException {
+        Path tmpSSTable = basePath.resolve(FILE_NAME + 0 + FILE_EXTENSION + TMP_FILE_EXTENSION);
+        Files.deleteIfExists(tmpSSTable);
+        Files.createFile(tmpSSTable);
+        EntriesMetadata entriesMetadata = data.get().countEntities();
+        return save(data, entriesMetadata.count(), entriesMetadata.entriesDataSize(), tmpSSTable);
+    }
+
+    public static List<SSTable> replaceSSTablesWithCompacted(
+            Arena arena,
+            Path compactedSSTable,
+            Path basePath,
+            List<SSTable> oldSSTables
+    ) throws IOException {
+        deleteOldSSTables(basePath, oldSSTables);
```

Comment:
если после этой строчки приложение упадет, то мы вообще все потеряем

Reply:
@pashchenko8 Не согласен - у нас останется FILE_NAME + 0 + FILE_EXTENSION + TMP_FILE_EXTENSION (см. [тут](https://github.com/polis-vk/2023-nosql-lsm/pull/212/files#diff-b09a55fbd5a2ea7d28e29d9fa17f12c035b5adad8868ac3481d50ef1107caf0bR232)).

Это скомпакченный файл. Стоило конечно отдельно называть его, но упустил этот момент. Однако в этом стейдже у нас нет требований по многопоточке.

Тогда при переоткрытии дао у нас вызовется метод checkIfCompactedExists, который проверит, что у нас есть на диске этот файл и что он корректно записан. Возьмёт удалит старые sstables (если они есть) и перезапишет FILE_NAME + 0 + FILE_EXTENSION + TMP_FILE_EXTENSION в FILE_NAME + 0 + FILE_EXTENSION

Reply:
обычно tmp File полностью игнорируются на старте приложения. поэтому я эту часть не заметил.
но давайте разберем все-же то по какому пути вы решили пойти:

checkIfCompactedExists не проверяет целостность файла... то-есть приложение могло упасть в момент когда вы заполняли tmpFile, потом в момент старта оно стартует, смотрит что файл есть и что просто у него есть начальный байт, после чего удаляются ваши старые sstable.
а то что было в tmp вообще сейчас не валидно. И мы имеем опять потерю данных

Reply:
@pashchenko8 Да, но я принудительно использую метод MemorySegment.force() https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/foreign/MemorySegment.html#force()

Если я правильно интерпретировал документацию, то он _гарантирует_ запись данных на диск

> when this method returns it is guaranteed that all changes made to this segment since it was created, or since this method was last invoked, will have been written to that device.

Использую я его тут - https://github.com/polis-vk/2023-nosql-lsm/pull/212/files#diff-b09a55fbd5a2ea7d28e29d9fa17f12c035b5adad8868ac3481d50ef1107caf0bR199

и тут - https://github.com/polis-vk/2023-nosql-lsm/pull/212/files#diff-b09a55fbd5a2ea7d28e29d9fa17f12c035b5adad8868ac3481d50ef1107caf0bR201

Mapping mode я использую READ_WRITE https://github.com/polis-vk/2023-nosql-lsm/pull/212/files#diff-b09a55fbd5a2ea7d28e29d9fa17f12c035b5adad8868ac3481d50ef1107caf0bR172

Reply:
В таком случае просмотр начального байта должен гарантировать целостность